_posts/2024-08-27-cheri-myths-safe-languages.markdown

```diff
@@ -0,0 +1,156 @@
+
+This was one of the problems that CHERI aimed to solve, in two ways.
+At the fine granularity, CHERI offers language-agnostic memory safety primitives.
+Compilers can generate, or humans can write, assembler making use of these primitives to capture and then enforce memory safety of source-level objects.
+The switcher is the most privileged part of a CHERIoT system and is written in assembly (it has to do things like save all registers, which are not possible in any high-level language).
+In spite of its privilege, it remains bound by the memory-safety rules.
+It cannot access any compartments other than the caller and callee.
+It cannot access any memory other than that reachable from the registers when it is invoked and its internal state, unlike a traditional OS kernel that can access anything.
+
+Enforcing these guarantees in the hardware makes it possible for safe-language code to call unsafe or, most excitingly, _differently_ safe code and yet still enforce source-level properties.
+Consider a single program composed of code written in all of Rust, Java, and Haskell.
+Each is in the 'safe languages' category, but provides different guarantees.
+Rust code passing a borrowed reference to Java may wish to ensure that it isn't captured, a property that CHERIoT can enforce directly (or CHERI can enforce with indirection).
+Haskell passing an object to Java or Rust may wish to guarantee deep immutability, a property that CHERIoT and all Morello or later CHERI platforms can enforce.
+
+At a coarser granularity, CHERI makes it easy to sandbox entire unsafe components.
+You can take a C library and put it in an isolated compartment.
+This is closely related to the fine-grained guarantees because, even if the unsafe library has a private heap, you are going to want to share some objects with it.
+Compression libraries are among the simplest to sandbox: they have a simple interface that takes an input buffer and provides an output buffer.
+CHERI makes it easy to provide a read-only view of an input buffer (avoiding the need to copy) and a write-only output buffer.
+
+In the CHERIoT network stack, we take advantage of this between the TLS and TCP/IP compartments.
+The TLS stack maintains its own ring buffers for messages.
+When it calls into the TCP/IP stack to receive some data, it passes a bounded capability to a subset of the buffer, with write-only, no-capture permissions.
+By the end of the call, the TLS stack knows that the TCP/IP stack doesn't have access to the buffer and can't have leaked any stale data in there.
+The TLS compartment contains BearSSL, which is mostly written in a domain-specific language that compiles down to C for side-channel-resistant crypto routines, but the same guarantees are available to any language that can target the platform.
+
+# Rewriting has an opportunity cost, if it's possible at all
+
+There's a pervasive narrative in some camps that rewriting everything in a safe language is the path to fixing security.
```

Comment:
I kinda think this section wants to have four sub-sections?
* Expense
* New bugs in rewrites
* Desirability of rewrites ("`ffmpeg` sure is fast; do you want to do all that again?")
* Feasibility of rewrites (TCB intrinsic unsafety)